pkg/translate/translateValue.go

```diff
@@ -0,0 +1,554 @@
+			return err
+		}
+
+		if _, err := DeleteFromTree(valueTree, util.ToYAMLPath(inPath), util.ToYAMLPath(inPath)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// translateRemainPaths translates remaining paths that are not availing in existing mappings.
+func (t *ValueYAMLTranslator) translateRemainPaths(valueTree map[string]interface{},
+	cpSpecTree map[string]interface{}, path util.Path) error {
+	for key, val := range valueTree {
+		newPath := append(path, key)
+		// value set to nil means no translation needed or being translated already.
+		if val == nil {
+			continue
+		} else {
+			switch node := val.(type) {
+			case map[string]interface{}:
+				err := t.translateRemainPaths(node, cpSpecTree, newPath)
+				if err != nil {
+					return err
+				}
+			case []interface{}:
+				errs := util.Errors{}
+				for _, newNode := range node {
+					newMap, ok := newNode.(map[string]interface{})
+					if !ok {
+						return fmt.Errorf("fail to convert node to map[string] interface")
+					}
+					errs = util.AppendErr(errs, t.translateRemainPaths(newMap, cpSpecTree, newPath))
+				}
+				if errs != nil {
+					return errs
+				}
+			// remaining leaf need to be put into common.values
+			default:
+				err := t.translateToCommonValues(cpSpecTree, newPath, val)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// DeleteFromTree sets value at path of input tree to nil
+func DeleteFromTree(valueTree map[string]interface{}, path util.Path, remainPath util.Path) (bool, error) {
+	if len(remainPath) == 0 {
+		return false, nil
+	}
+	for key, val := range valueTree {
+		if key == remainPath[0] {
+			// found the path to delete value
+			if len(remainPath) == 1 {
+				valueTree[key] = nil
+				return true, nil
+			}
+			remainPath = remainPath[1:]
+			switch node := val.(type) {
+			case map[string]interface{}:
+				return DeleteFromTree(node, path, remainPath)
+			case []interface{}:
+				for _, newNode := range node {
+					newMap, ok := newNode.(map[string]interface{})
+					if !ok {
+						return false, fmt.Errorf("fail to translate to map[string]interface{}")
+					}
+					found, err := DeleteFromTree(newMap, path, remainPath)
+					if found && err == nil {
+						return found, nil
+					}
+				}
+			// leaf
+			default:
+				return false, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+// translateToCommonValues translates paths to common.values
+func (t *ValueYAMLTranslator) translateToCommonValues(root map[string]interface{}, path util.Path, value interface{}) error {
+	if len(path) == 0 {
+		log.Warn("path is empty when translating to common.values")
+		return nil
+	}
+	// path starts with ValueComponent name or global
+	cn := path[0]
+	newCN := t.ValuesToComponentName[cn]
+	if newCN == "" {
+		// some components path len is 2, e.g. mixer.policy
+		if len(path) > 1 {
+			cn += path[1]
+			newCN = t.ValuesToComponentName[cn]
+		}
+		if newCN == "" {
+			return nil
+		}
+	}
+	// skip enablement node
```

Comment:
mention why you're skipping - in general in comments try to avoid stating what the code is doing if it's fairly obvious and write about why it's doing it. usually if code is well written, the "what it's doing" should be self-documenting. if that's not the case it can signal the need to refactor.

Reply:
done